### PR TITLE
"My Libraries" tab - unpublish libraries

### PR DESCRIPTION
**Pull request by maddiedierker**
Follow-up to #33329 and #33330

Adds a "Quick Actions" column to the "My Libraries" table to allow users to unpublish libraries from `/projects/libraries`:

Screen Shot 2020-02-28 at 1 30 03 PM

[Video of current unpublish flow](https://slack-files.com/T039SAH7W-FUNPEFA94-7857932767)

## Links

- [STAR-960](https://codedotorg.atlassian.net/browse/STAR-960)
- [STAR-748](https://codedotorg.atlassian.net/browse/STAR-748)
- UPCOMING: [STAR-992](https://codedotorg.atlassian.net/browse/STAR-992) (improvements to the unpublish experience -- a confirmation modal + error handling)

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by jmkulwik** [comment]
One thought for the unpublish flow - This seems like an action we'd want to have a confirmation dialog for. Or maybe a half-way step would be to make the button gray instead of orange.

File: apps/src/code-studio/components/libraries/LibraryClientApi.js [hunk 1]
```diff
@@ -25,6 +26,22 @@ export default class LibraryClientApi {
     );
   }
 
+  unpublish(project, callback) {
```
Comment (jmkulwik) [inline apps/src/code-studio/components/libraries/LibraryClientApi.js:29]: It's a bit confusing as to when to use `unpublish` vs `delete`. Would it make sense to make `unpublish` the default action, have it check whether we are in an app, and then take actions accordingly? I'm thinking about here: https://github.com/code-dot-org/code-dot-org/blob/c09f509ce6d5ca5af6effb6023066374d091c95c/apps/src/code-studio/components/libraries/LibraryPublisher.jsx#L234-L246

Another possibility is to add some comments to make it clear when to use each.

File: apps/src/templates/projects/LibraryTable.jsx [hunk 2]
```diff
@@ -188,16 +193,53 @@ class LibraryTable extends React.Component {
...
+    );
+  };
+
   render() {
     if (!this.props.personalProjectsList) {
       // Projects haven't loaded from server yet, so display nothing.
       return null;
     }
 
+    const libraries = _.filter(
```
Comment (jmkulwik) [inline apps/src/templates/projects/LibraryTable.jsx]: Rather than using lodash, you should be able to just do this to keep things clearer:
`const libraries = this.props.personalProjectsList.filter(p => p.libraryName)`

File: apps/src/templates/projects/projectsRedux.js [hunk 3]
```diff
@@ -193,6 +200,17 @@ function personalProjectsList(state = initialPersonalProjectsList, action) {
         ...state,
         projects: action.personalProjectsList
       };
+    case UPDATE_PERSONAL_PROJECT_DATA:
+      var projectsList = [...state.projects];
+      var updatedProjectIdx = projectsList.findIndex(
```
Comment (jmkulwik) [inline apps/src/templates/projects/projectsRedux.js]: NIT: `updatedProjectIdx` should be `updatedProjectIndex`

File: apps/src/templates/projects/projectsRedux.js [hunk 4]
```diff
@@ -193,6 +200,17 @@ function personalProjectsList(state = initialPersonalProjectsList, action) {
...
+        p => p.channel === action.projectId
```
Comment (jmkulwik) [inline apps/src/templates/projects/projectsRedux.js]: NIT: `p` should be `project`

File: apps/src/templates/projects/LibraryTable.jsx [hunk 5]
```diff
@@ -188,16 +193,53 @@ class LibraryTable extends React.Component {
...
+      this.props.personalProjectsList,
+      p => p.libraryName
```
Comment (jmkulwik) [inline apps/src/templates/projects/LibraryTable.jsx]: NIT: `p` should be `project`

File: apps/src/templates/projects/projectsRedux.js [hunk 6]
```diff
@@ -463,6 +469,27 @@ export function unpublishProject(projectId) {
...
+  projectId,
+  libraryApi = new LibraryClientApi(projectId),
+  onComplete = () => {}
+) => {
+  return dispatch => {
+    fetchProjectToUpdate(projectId, (error, data) => {
+      if (error) {
+        onComplete(error, data);
+      } else {
+        libraryApi.unpublish(data, (error, serverData) => {
```
Comment (jmkulwik) [inline apps/src/templates/projects/projectsRedux.js:482]: I think you can just call unpublish. There _should_ be different error codes if the project just doesn't exist (404) vs if there is a server error.

File: apps/test/unit/code-studio/components/libraries/LibraryClientApiTest.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
+import {assert} from '../../../../util/reconfiguredChai';
+import sinon from 'sinon';
+import LibraryClientApi from '@cdo/apps/code-studio/components/libraries/LibraryClientApi';
+
+describe('LibraryClientApi', () => {
```
Comment (jmkulwik) [inline apps/test/unit/code-studio/components/libraries/LibraryClientApiTest.js:5]: Thank you for adding these! 🎉

File: shared/middleware/channels_api.rb [hunk 8]
```diff
@@ -155,7 +155,10 @@ class ChannelsApi < Sinatra::Base
     end
     bad_request unless value.is_a? Hash
     value = value.merge('updatedAt' => Time.now)
-    value = value.merge('libraryPublishedAt' => Time.now) if value['publishLibrary']
+
+    # Set libraryPublishedAt timestamp if we are publishing a project library.
+    publish_library = value.delete('publishLibrary')
```
Comment (jmkulwik) [inline shared/middleware/channels_api.rb:160]: What does value.delete do here? (for my own learning)

Comment (maddiedierker) [reply shared/middleware/channels_api.rb:160]: it removes `publishLibrary` from the `value` hash, which is only superficially useful to us -- we overwrite the entire project in `PEGASUS_DB[:storage_apps]` with the `value` hash, and it felt gross to write `publishLibrary` to the db

so there wasn't any user effect, but it would start gunking up our storage_apps table

File: apps/src/templates/projects/projectsRedux.js (hunk 6 above)
Comment (maddiedierker) [reply apps/src/templates/projects/projectsRedux.js:482]: i'm not sure what you mean -- can you rephrase that?

File: apps/src/code-studio/components/libraries/LibraryClientApi.js (hunk 1 above)
Comment (maddiedierker) [reply apps/src/code-studio/components/libraries/LibraryClientApi.js:29]: i think a comment would be more appropriate -- `delete` only removes the asset from S3, so this method encompasses that `delete` action plus updating the pegasus db

Comment (maddiedierker) [reply apps/src/code-studio/components/libraries/LibraryClientApi.js:29]: ah, sorry, i think i misread your comment -- are you asking if we should update `<LibraryPublisher/>` to use this new `unpublish` method instead of the existing `delete` method?

Comment (jmkulwik) [reply apps/src/code-studio/components/libraries/LibraryClientApi.js:29]: Yeah, sorry that wasn't very clear. Your second comment is what I meant.

File: apps/src/templates/projects/projectsRedux.js (hunk 6 above)
Comment (jmkulwik) [reply apps/src/templates/projects/projectsRedux.js:482]: Nevermind. I had the wrong thing in mind for what fetchProjectToUpdate does.

File: shared/middleware/channels_api.rb (hunk 8 above)
Comment (jmkulwik) [reply shared/middleware/channels_api.rb:160]: Ok. Makes sense. Thank you!

File: apps/src/code-studio/components/libraries/LibraryClientApi.js (hunk 1 above)
Comment (maddiedierker) [reply apps/src/code-studio/components/libraries/LibraryClientApi.js:29]: all good! i'll go ahead and update `<LibraryPublisher/>` now so i don't forget 😅

Comment (jmkulwik) [reply apps/src/code-studio/components/libraries/LibraryClientApi.js:29]: Thanks for making this change!

**Review by jmkulwik — APPROVED** [review]
